### PR TITLE
Add initial Fuse implementation for Stream

### PR DESCRIPTION
**Pull request by spacejam**
@matklad does this address your use case?

### CODE REVIEW

File: src/stream/mod.rs [hunk 1]
```diff
@@ -25,7 +25,7 @@
 pub use empty::{empty, Empty};
 pub use once::{once, Once};
 pub use repeat::{repeat, Repeat};
-pub use stream::{Stream, Take};
+pub use stream::{Fuse, Stream, Take};
```
Comment (yoshuawuyts) [inline src/stream/mod.rs]: Sticking with the `std` shouldn't this be `async_std::stream::FusedStream` as a counterpart to `std::iter::FusedIterator`? I know we haven't been really naming things that way, but figured it'd be worth bringing up.

cc/ @stjepang thoughts on naming?

Comment (skade) [reply src/stream/mod.rs]: It should.

Comment (matklad) [reply src/stream/mod.rs]: FusedIterator is a trait,the struct is named Fuse: https://doc.rust-lang.org/std/iter/struct.Fuse.html.

Comment (yoshuawuyts) [reply src/stream/mod.rs]: Ah my bad, thanks for catching that @matklad!

File: src/stream/stream.rs [hunk 2]
```diff
@@ -171,3 +203,54 @@ impl<S: futures::Stream> futures::Stream for Take<S> {
...
+    pub fn is_done(&self) -> bool {
+        self.done
+    }
+
+    /// Consumes this `Fuse` and returns the inner
+    /// `Stream`, unfusing it if it had become
+    /// fused.
+    pub fn into_inner(self) -> S
+    where
+        S: Sized,
```
[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply src/stream/stream.rs]: Added back the `where Self: Sized` bound in the `fn fused()` impl. I couldn't get this to compile with `S: ?Sized` in the impl bound.

File: src/stream/stream.rs [hunk 3]
```diff
@@ -114,6 +114,38 @@ pub trait Stream {
...
+    /// # fn main() { async_std::task::block_on(async {
+    /// #
+    /// use async_std::prelude::*;
+    /// use async_std::stream;
+    ///
+    /// let mut s = stream::repeat(9).take(3);
+    ///
+    /// while let Some(v) = s.next().await {
+    ///     assert_eq!(v, 9);
+    /// }
```
[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply src/stream/stream.rs]: I've updated the example!

**Comment by yoshuawuyts** [comment]
I've rebased @spacejam's excellent work on master; it's all passing now. Also removed the two helper methods on the `Fuse` struct that don't have counterparts in std so we only add the minimal API surface for now.

Given the amount of review this has had, I think this should be good to merge. Ref #129. 

bors r+

**Pull request merged**